### PR TITLE
Ensure GCHeap related debugging still works even when we use newer CLRGC to target older runtimes

### PR DESCRIPTION
**Pull request by cshung**
This change makes sure GCHeap-related debugging still works even when we use newer CLRGC to target older runtimes by:

- Placing all the newer fields at the end of `GcDacVars` and `dac_gcheap`, and
- Making sure GC never populate fields that do not exist on older versions.

### CODE REVIEW

File: src/coreclr/gc/gcinterface.dacvars.def [hunk 1]
```diff
@@ -68,6 +79,11 @@ GC_DAC_PTR_VAR    (uint8_t,               bookkeeping_start)
 GC_DAC_ARRAY_VAR  (dac_region_free_list,  global_regions_to_decommit)
 GC_DAC_PTR_VAR    (dac_region_free_list,  global_free_huge_regions)
 GC_DAC_ARRAY_VAR  (dac_region_free_list,  free_regions)
+GC_DAC_PTR_VAR    (dac_heap_segment,      freeable_soh_segment)
```
Comment (noahfalk) [inline src/coreclr/gc/gcinterface.dacvars.def:82]: Is this the start of the minor version 2 fields? I'd suggest inserting a comment saying that. It helps document how the contract changed in each version.

**Review by noahfalk — APPROVED** [review]
A few suggestions inline but LGTM.

fyi @dotnet/dotnet-diag

File: src/coreclr/gc/dac_gcheap_fields.h [hunk 2]
```diff
@@ -30,14 +37,30 @@ DEFINE_MISSING_FIELD(mark_array)
...
-DEFINE_MISSING_FIELD(freeable_soh_segment)
-DEFINE_MISSING_FIELD(freeable_uoh_segment)
 DEFINE_MISSING_FIELD(saved_sweep_ephemeral_seg)
 DEFINE_MISSING_FIELD(saved_sweep_ephemeral_start)
 #endif // defined(ALL_FIELDS) || defined(BACKGROUND_GC)
 
+// generation_table is a special field, it is not included in dac_gcheap
+// but its offset need to be right at this spot in order to match the 
+// layout we had in .NET 6+
+DEFINE_OFFSET_ONLY_FIELD(generation_table)
```
[the remark this answers is not in the document]
Comment (cshung) [reply src/coreclr/gc/dac_gcheap_fields.h]: That's a good idea. I pushed a change that implement the idea, testing it out now.

**Comment by mikem8361** [comment]
Is this going to be merged? It would be nice if .NET 8 doesn't have this problem going forward.

**Review by Maoni0 — APPROVED** [review]
:shipit:

**Pull request merged**